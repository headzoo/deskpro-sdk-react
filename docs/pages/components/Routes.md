Overview
========
The SDK includes a light-weight router which displays different pages within DeskPRO apps. Route locations are defined using simple names instead of addresses, because DeskPRO apps do not use the browser location (URL). The current location is changed by calling [this.props.route.to()](/pages/props/#route) with the name of the new location, or by using the [Link](/pages/components/Link) and [LinkButton](/pages/components/LinkButton) components.

## Example
First create a few pages for the app.

```jsx
// PageSettings.jsx
import React from 'react';
import { Link, sdkConnect } from 'deskpro-sdk-react';

const PageSettings = () => (
    <div>
        <h1>Settings</h1>
        <Link to="index">
            Go to index
        </Link>
    </div>
);

export default sdkConnect(PageSettings);
```

```jsx
// PageIndex.jsx
import React from 'react';
import { Link, sdkConnect } from 'deskpro-sdk-react';

const PageSettings = () => (
    <div>
        <h1>Index</h1>
        <Link to="settings">
            Go to settings
        </Link>
    </div>
);

export default sdkConnect(PageSettings);
```

```jsx
// PageNotFound.jsx
import React from 'react';

const PageNotFound = () => (
    <div>
        <h1>Not Found</h1>
    </div>
);

export default PageNotFound;
```

Now create the main app component which contains the `Routes`. The `Route` which matches the current location gets rendered. The route with the `defaultRoute` prop renders when no other routes matches the current location.

```jsx
// App.jsx
import React from 'react';
import { Routes, Route, sdkConnect } from 'deskpro-sdk-react';
import PageSettings from './PageSettings';
import PageIndex from './PageIndex';
import PageNotFound from './PageNotFound';

class App extends React.Component {
  render() {
    return (
      <Routes>
        <Route location="settings" component={PageSettings} />
        <Route location="index" component={PageIndex} />
        <Route defaultRoute component={PageNotFound} />
      </Routes>
    );
  }
}

export default sdkConnect(App);
```

Routes may also be defined using child components instead of the `component` prop.

```jsx
<Routes>
    <Route location="settings">
        <PageSettings />
    </Route>
    <Route location="index">
        <PageIndex />
    </Route>
    <Route defaultRoute>
        <h1>Not Found</h1>
    </Route>
</Routes>
```

## Changing routes manually

Call `this.props.route.to()` to change routes manually.

```jsx
import React from 'react';
import { sdkConnect } from 'deskpro-sdk-react';

class PageSettings extends Route.Component {
    handleClick = () => {
        this.props.route.to('index');
    };
    
    render() {
        return (
            <div>
                <h1>Settings</h1>
                <button onClick={this.handleClick}>
                    Go to index
                </button>
            </div>
        );
    }
};

export default sdkConnect(PageSettings);
```

## See also
* [this.props.route](/pages/props/#route)
* [Link](/pages/components/Link)
* [LinkButton](/pages/components/LinkButton)
