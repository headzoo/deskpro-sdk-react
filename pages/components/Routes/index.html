<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">   
    <link rel="shortcut icon" href="../../../favicon.png">

    <title>Routes - DeskPRO React SDK</title>

    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css">
    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/font-hack/2.018/css/hack.min.css">
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto">
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css">
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.5.0/styles/agate.min.css">
    <link rel="stylesheet" href="../../../css/base.css">
    <link rel="stylesheet" href="../../../css/psinder.css">
    
    


    <link href="../../../css/site.css" rel="stylesheet">


    <link href="../../../css/admonition.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="//oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="//oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

    
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="../../..">
                <img src="../../../logo.png" />
                DeskPRO React SDK
            </a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="../../..">Overview</a>
                    </li>
                
                
                
                    <li >
                        <a href="../../manifest/">Manifest</a>
                    </li>
                
                
                
                    <li >
                        <a href="../../props/">Props</a>
                    </li>
                
                
                
                    <li class="dropdown active">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Components <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../DeskproSDK/">DeskproSDK</a>
</li>

                        
                            
<li class="active">
    <a href="./">Routes</a>
</li>

                        
                            
<li >
    <a href="../Link/">Link</a>
</li>

                        
                            
<li >
    <a href="../LinkButton/">LinkButton</a>
</li>

                        
                        </ul>
                    </li>
                
                
                
                    <li class="dropdown">
                        <a href="#" class="dropdown-toggle" data-toggle="dropdown">Tutorials <b class="caret"></b></a>
                        <ul class="dropdown-menu">
                        
                            
<li >
    <a href="../../tutorials/basic/">Basic App</a>
</li>

                        
                            
<li >
    <a href="../../tutorials/form/">Form Handling</a>
</li>

                        
                            
<li >
    <a href="../../tutorials/oauth/">Using OAuth</a>
</li>

                        
                            
<li >
    <a href="../../tutorials/redux/">Using Redux</a>
</li>

                        
                        </ul>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li>
                        <a href="https://github.com/DeskproApps/sdk-react">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="first-level active"><a href="#overview">Overview</a></li>
        
            <li class="second-level"><a href="#example">Example</a></li>
            
        
            <li class="second-level"><a href="#changing-routes-manually">Changing routes manually</a></li>
            
        
            <li class="second-level"><a href="#see-also">See also</a></li>
            
        
    
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<h1 id="overview">Overview</h1>
<p>The SDK includes a light-weight router which displays different pages within DeskPRO apps. Route locations are defined using simple names instead of addresses, because DeskPRO apps do not use the browser location (URL). The current location is changed by calling <a href="../../../pages/props/#route">this.props.route.to()</a> with the name of the new location, or by using the <a href="../../../pages/components/Link">Link</a> and <a href="../../../pages/components/LinkButton">LinkButton</a> components.</p>
<h2 id="example">Example</h2>
<p>First create a few pages for the app.</p>
<pre><code class="jsx">// PageSettings.jsx
import React from 'react';
import { Link, sdkConnect } from 'deskpro-sdk-react';

const PageSettings = () =&gt; (
    &lt;div&gt;
        &lt;h1&gt;Settings&lt;/h1&gt;
        &lt;Link to=&quot;index&quot;&gt;
            Go to index
        &lt;/Link&gt;
    &lt;/div&gt;
);

export default sdkConnect(PageSettings);
</code></pre>

<pre><code class="jsx">// PageIndex.jsx
import React from 'react';
import { Link, sdkConnect } from 'deskpro-sdk-react';

const PageSettings = () =&gt; (
    &lt;div&gt;
        &lt;h1&gt;Index&lt;/h1&gt;
        &lt;Link to=&quot;settings&quot;&gt;
            Go to settings
        &lt;/Link&gt;
    &lt;/div&gt;
);

export default sdkConnect(PageSettings);
</code></pre>

<pre><code class="jsx">// PageNotFound.jsx
import React from 'react';

const PageNotFound = () =&gt; (
    &lt;div&gt;
        &lt;h1&gt;Not Found&lt;/h1&gt;
    &lt;/div&gt;
);

export default PageNotFound;
</code></pre>

<p>Now create the main app component which contains the <code>Routes</code>. The <code>Route</code> which matches the current location gets rendered. The route with the <code>defaultRoute</code> prop renders when no other routes matches the current location.</p>
<pre><code class="jsx">// App.jsx
import React from 'react';
import { Routes, Route, sdkConnect } from 'deskpro-sdk-react';
import PageSettings from './PageSettings';
import PageIndex from './PageIndex';
import PageNotFound from './PageNotFound';

class App extends React.Component {
  render() {
    return (
      &lt;Routes&gt;
        &lt;Route location=&quot;settings&quot; component={PageSettings} /&gt;
        &lt;Route location=&quot;index&quot; component={PageIndex} /&gt;
        &lt;Route defaultRoute component={PageNotFound} /&gt;
      &lt;/Routes&gt;
    );
  }
}

export default sdkConnect(App);
</code></pre>

<p>Routes may also be defined using child components instead of the <code>component</code> prop.</p>
<pre><code class="jsx">&lt;Routes&gt;
    &lt;Route location=&quot;settings&quot;&gt;
        &lt;PageSettings /&gt;
    &lt;/Route&gt;
    &lt;Route location=&quot;index&quot;&gt;
        &lt;PageIndex /&gt;
    &lt;/Route&gt;
    &lt;Route defaultRoute&gt;
        &lt;h1&gt;Not Found&lt;/h1&gt;
    &lt;/Route&gt;
&lt;/Routes&gt;
</code></pre>

<h2 id="changing-routes-manually">Changing routes manually</h2>
<p>Call <code>this.props.route.to()</code> to change routes manually.</p>
<pre><code class="jsx">import React from 'react';
import { sdkConnect } from 'deskpro-sdk-react';

class PageSettings extends Route.Component {
    handleClick = () =&gt; {
        this.props.route.to('index');
    };

    render() {
        return (
            &lt;div&gt;
                &lt;h1&gt;Settings&lt;/h1&gt;
                &lt;button onClick={this.handleClick}&gt;
                    Go to index
                &lt;/button&gt;
            &lt;/div&gt;
        );
    }
};

export default sdkConnect(PageSettings);
</code></pre>

<h2 id="see-also">See also</h2>
<ul>
<li><a href="../../../pages/props/#route">this.props.route</a></li>
<li><a href="../../../pages/components/Link">Link</a></li>
<li><a href="../../../pages/components/LinkButton">LinkButton</a></li>
</ul></div>
        
    </div>

    <footer class="col-md-12 text-center">
        
            
            <p><a href="https://github.com/DeskproApps/sdk-react/tree/master/docs/pages/components/Routes.md" class="icon icon-github"> Edit this page on GitHub</a></p>
            
        
        
            <small>Copyright 2001 – 2017 Deskpro Ltd. All rights reserved.<br></small>
        
    </footer>

    <script src="//code.jquery.com/jquery-2.2.4.min.js"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.0.3/js/bootstrap.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
    <script>
    var base_url = '../../..';
    </script>
    <script data-main="../../../mkdocs/js/search.js" src="../../../mkdocs/js/require.js"></script>
    <script src="../../../js/base.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span>
                        <span class="sr-only">Close</span>
                    </button>
                    <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                </div>
                <div class="modal-body">
                    <p>
                        From here you can search these documents. Enter your search terms below.
                    </p>
                    <form role="form">
                        <div class="form-group">
                            <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                        </div>
                    </form>
                    <div id="mkdocs-search-results"></div>
                </div>
            </div>
        </div>
    </div>
    </body>
</html>
