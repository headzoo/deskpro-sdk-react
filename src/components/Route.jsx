import React from 'react';
import PropTypes from 'prop-types';

/**
 * Used by the Routes component to match components to route locations.
 *
 * Example:
 *
 * ```jsx
 * <Routes>
 *  <Route location="settings" component={PageSettings} />
 *  <Route location="index" component={PageIndex} />
 *  <Route defaultRoute component={PageNotFound} />
 * </Routes>
 * ```
 */
export default class Route extends React.Component {
  static propTypes = {
    location:     PropTypes.string,
    defaultRoute: PropTypes.bool,
    component:    PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    children:     PropTypes.node
  };

  static defaultProps = {
    location:     '',
    children:     '',
    component:    '',
    defaultRoute: false
  };

  /**
   * @returns {string}
   */
  location = () => {
    return this.props.location;
  };

  /**
   * @returns {React.Component}
   */
  component = () => {
    return this.props.component;
  };

  /**
   * @returns {boolean}
   */
  defaultRoute = () => {
    return this.props.defaultRoute;
  };

  /**
   * @returns {XML}
   */
  render() {
    if (this.props.component) {
      return React.createElement(this.props.component);
    }
    return this.props.children;
  }
}
