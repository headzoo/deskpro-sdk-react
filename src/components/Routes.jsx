import React from 'react';
import PropTypes from 'prop-types';
import Route from './Route';

/**
 * Ensures the Routes children are of type Route
 *
 * @returns {*}
 */
const childrenPropType = (props, propName, componentName) => {
  const prop = props[propName];
  let error  = null;
  React.Children.forEach(prop, (child) => {
    if (child.type !== Route) {
      error = new Error(`${componentName} children should be of type 'Route'`);
    }
  });

  return error;
};

/**
 * Renders the child which matches the current route.
 *
 * Example:
 *
 * ```jsx
 * <Routes>
 *  <Route location="settings" component={PageSettings} />
 *  <Route location="index" component={PageIndex} />
 *  <Route defaultRoute component={PageNotFound} />
 * </Routes>
 * ```
 */
export default class Routes extends React.Component {
  static propTypes = {
    /**
     * Name of the route to render. Overrides the location in the store.
     */
    to: PropTypes.string,

    /**
     * Child components.
     */
    children: childrenPropType
  };

  /**
   * Specifies the default values for props
   */
  static defaultProps = {
    to:       '',
    children: []
  };

  /**
   * Context values the child wants passed down from the parent
   */
  static contextTypes = {
    route: PropTypes.object
  };

  /**
   * Invoked immediately before mounting occurs
   */
  componentDidMount() {
    if (this.props.to === '' && this.context.route === undefined) {
      throw new Error('Route context not found. Did you forget to use sdkConnect?');
    }
  }

  /**
   * Creates or clones the component stored in the given route
   *
   * @param {Route} route
   * @returns {XML}
   */
  createRouteComponent = (route) => {
    if (route.props.component) {
      return React.createElement(route.props.component);
    }
    return React.cloneElement(React.Children.only(route.props.children));
  };

  /**
   * @returns {XML}
   */
  render() {
    const location = this.props.to || this.context.route.location;
    const routes   = React.Children.toArray(this.props.children);

    let defaultRoute = null;
    for (let i = 0; i < routes.length; i++) {
      if (routes[i].props.location === location) {
        return this.createRouteComponent(routes[i]);
      }
      if (routes[i].props.defaultRoute) {
        defaultRoute = routes[i];
      }
    }
    if (defaultRoute) {
      return this.createRouteComponent(defaultRoute);
    }

    throw new Error(`Route not found for location '${location}' and defaultRoute not found.`);
  }
}
