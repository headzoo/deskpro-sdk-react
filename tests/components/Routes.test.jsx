import React from 'react';
import renderer from 'react-test-renderer';
import Routes from '../../src/components/Routes';
import Route from '../../src/components/Route';

it('Routes renders the matching component', () => {
  const Page1 = () => <div>Page1</div>;
  const Page2 = () => <div>Page2</div>;

  const renderedValue = renderer.create(
    <Routes to="page2">
      <Route location="page1" component={Page1} />
      <Route location="page2" component={Page2} />
    </Routes>
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});

it('Routes renders the defaultRoute', () => {
  const Page1 = () => <div>Page1</div>;
  const Page2 = () => <div>Page2</div>;

  const renderedValue = renderer.create(
    <Routes to="page3">
      <Route defaultRoute component={Page1} />
      <Route location="page2" component={Page2} />
    </Routes>
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
