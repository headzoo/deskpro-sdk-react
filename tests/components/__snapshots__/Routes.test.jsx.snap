// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes renders the defaultRoute 1`] = `
<div>
  Page1
</div>
`;

exports[`Routes renders the matching component 1`] = `
<div>
  Page2
</div>
`;
